### PR TITLE
Add sinuosity expression

### PR DESCRIPTION
**Pull request by lbartoletti**
## Description

This PR adds the sinuosity function in the expressions.

### CODE REVIEW

File: src/core/expression/qgsexpressionfunction.cpp [hunk 1]
```diff
@@ -3572,6 +3572,19 @@ static QVariant fcnMMax( const QVariantList &values, const QgsExpressionContext
   return QVariant( max );
 }
 
+static QVariant fcnSinuosity( const QVariantList &values, const QgsExpressionContext *, QgsExpression *parent, const QgsExpressionNodeFunction * )
+{
+  QgsGeometry geom = QgsExpressionUtils::getGeometry( values.at( 0 ), parent );
+  const QgsCurve *curve = qgsgeometry_cast< const QgsCurve * >( geom.constGet() );
+  if ( ( geom.type() != QgsWkbTypes::LineGeometry ) || !curve )
```
Comment (nyalldawson) [inline src/core/expression/qgsexpressionfunction.cpp:3579]:
```suggestion
  if ( !curve )
```

The first check is redundant

File: tests/src/core/testqgsexpression.cpp [hunk 2]
```diff
@@ -1316,6 +1316,12 @@ class TestQgsExpression: public QObject
       QTest::newRow( "main angle line" ) << "round(main_angle( geom_from_wkt('LINESTRING (-1 2, 9 12)') ))" << false << QVariant( 45 );
       QTest::newRow( "main angle not geom" ) << "main_angle('g')" << true << QVariant();
       QTest::newRow( "main angle null" ) << "main_angle(NULL)" << false << QVariant();
+      QTest::newRow( "sinuosity not geom" ) << "sinuosity('g')" << true << QVariant();
+      QTest::newRow( "sinuosity null" ) << "sinuosity(NULL)" << false << QVariant();
+      QTest::newRow( "sinuosity point" ) << "sinuosity(geom_from_wkt('POINT(1 2)'))" << true << QVariant( );
+      QTest::newRow( "sinuosity multi linestring" ) << "sinuosity(geom_from_wkt('MULTILINESTRING( (0 0, 1 1), (2 2, 3 3) )'))" << true << QVariant( );
+      QTest::newRow( "sinuosity linestring" ) << "round(sinuosity(geom_from_wkt(' LINESTRING(2 0, 2 2, 3 2, 3 3)')), 3)" << false << QVariant( 1.265 );
+      QTest::newRow( "sinuosity linestring" ) << "sinuosity(geom_from_wkt(' LINESTRING( 3 1, 5 1) '))" << false << QVariant( 1.0 );
```
Comment (nyalldawson) [inline tests/src/core/testqgsexpression.cpp:1324]: How about a test of a curved geometry too?

**Review by nyalldawson — COMMENTED** [review]
Nice!

**Review by DelazJ — COMMENTED** [review]
Just some spacing removal... I don't remember what is the practice here but do we indicate that it only works with single geometry (or it does not work with multi)?

**Comment by lbartoletti** [comment]
> I don't remember what is the practice here but do we indicate that it only works with single geometry

Yes, the algoritmh works only with single geometry.

File: src/core/expression/qgsexpressionfunction.cpp (hunk 1 above)
Comment (lbartoletti) [reply src/core/expression/qgsexpressionfunction.cpp:3579]: Indeed! thanks

File: tests/src/core/testqgsexpression.cpp (hunk 2 above)
Comment (lbartoletti) [reply tests/src/core/testqgsexpression.cpp]: Done and closed string.

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**